### PR TITLE
MT-76: merge opentxs changes part 1

### PR DESCRIPTION
**Pull request by radexx99**
New pool request with cleaned commits

### CODE REVIEW

File: generated/Version.hpp.in [hunk 1]
```diff
@@ -38,11 +38,15 @@ namespace session
 
 namespace blockchain
 {
-namespace block
-{
 namespace bitcoin
 {
+namespace block
```
Comment (TomaszAugustyn) [inline generated/Version.hpp.in:43]: For future improvements: is it necessary to keep in our repo generated files? The comment is general not specific to this file or line (I cannot find a way in Github to comment on the whole file)

**Review by bartalke — APPROVED** [review]

File: include/opentxs/blockchain/bitcoin/block/Header.hpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    auto as_Bitcoin() const noexcept -> const Header& final;
+    auto MerkleRoot() const noexcept -> const blockchain::block::Hash&;
+    auto Encode() const noexcept -> OTData;
+    auto Nonce() const noexcept -> std::uint32_t;
+    auto nBits() const noexcept -> std::uint32_t;
+    auto Timestamp() const noexcept -> Time;
+    auto Version() const noexcept -> std::uint32_t;
+
+    auto swap(Header& rhs) noexcept -> void;
```
Comment (TomaszAugustyn) [inline include/opentxs/blockchain/bitcoin/block/Header.hpp:30]: Inside one class we have mixed 3 naming conventions, e.g.:
- `as_Bitcoin` -> camel_Snake_Case
- `MerkleRoot` -> PascalCase
- `nBits`, `swap` -> camelCase

Can we please unify this?

File: include/opentxs/blockchain/block/Header.hpp [hunk 3]
```diff
@@ -46,63 +49,43 @@ namespace opentxs::blockchain::block
...
-    virtual auto ParentHash() const noexcept -> const block::Hash& = 0;
-    virtual auto ParentWork() const noexcept -> OTWork = 0;
-    virtual auto Position() const noexcept -> block::Position = 0;
-    virtual auto Print() const noexcept -> UnallocatedCString = 0;
-    OPENTXS_NO_EXPORT virtual auto Serialize(SerializedType& out) const noexcept
-        -> bool = 0;
-    virtual auto Serialize(
+        -> const blockchain::bitcoin::block::Header&;
+    auto clone() const noexcept -> std::unique_ptr<Header>;
+    auto Difficulty() const noexcept -> OTWork;
```
Comment (TomaszAugustyn) [inline include/opentxs/blockchain/block/Header.hpp:57]: same here - 3 different naming conventions

File: src/api/crypto/blockchain/Imp_blockchain.cpp [hunk 4]
```diff
@@ -381,72 +409,82 @@ auto BlockchainImp::ProcessMergedContact(
...
         }
 
-        copy.swap(tx);
-    } else {
-        if (false == db.StoreTransaction(in)) {
-            LogError()(OT_PRETTY_CLASS())("failed to save new transaction ")(
-                id.asHex())
+        if (false == db.AssociateTransaction(id, tx.Internal().GetPatterns())) {
+            LogError()(OT_PRETTY_CLASS())(
+                "associate patterns for transaction ")(id.asHex())
```
Comment (TomaszAugustyn) [inline src/api/crypto/blockchain/Imp_blockchain.cpp]: Shouldn't be "failed to associate patterns for transaction" ?

File: src/blockchain/bitcoin/block/header/Header.hpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+public:
+    virtual auto clone_bitcoin() const noexcept -> std::unique_ptr<Imp>
+    {
+        return std::make_unique<Imp>();
+    }
+
+    virtual auto MerkleRoot() const noexcept -> const blockchain::block::Hash&;
+    virtual auto Encode() const noexcept -> OTData;
+    virtual auto Nonce() const noexcept -> std::uint32_t { return {}; }
+    virtual auto nBits() const noexcept -> std::uint32_t { return {}; }
```
Comment (TomaszAugustyn) [inline src/blockchain/bitcoin/block/header/Header.hpp:70]: same here - 3 conventions

File: src/blockchain/bitcoin/block/header/Header.hpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+#include "blockchain/block/header/Imp.hpp"
+#include "internal/blockchain/bitcoin/block/Header.hpp"
+#include "internal/blockchain/block/Block.hpp"
+#include "opentxs/blockchain/BlockchainType.hpp"
+#include "opentxs/blockchain/Types.hpp"
+#include "opentxs/blockchain/bitcoin/NumericHash.hpp"
+#include "opentxs/blockchain/bitcoin/Work.hpp"
+#include "opentxs/blockchain/bitcoin/block/Header.hpp"
+#include "opentxs/blockchain/block/Hash.hpp"
+#include "opentxs/blockchain/block/Hash.hpp"
```
Comment (TomaszAugustyn) [inline src/blockchain/bitcoin/block/header/Header.hpp]: The same header added twice.

File: src/blockchain/bitcoin/block/header/Header.hpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+namespace block
+{
+class Hash;
+}  // namespace block
+}  // namespace blockchain
+// }  // namespace v1
+}  // namespace opentxs
+// NOLINTEND(modernize-concat-nested-namespaces)
+
+namespace be = boost::endian;
```
Comment (TomaszAugustyn) [inline src/blockchain/bitcoin/block/header/Header.hpp]: It's not used anywhere in the header.

File: src/blockchain/bitcoin/block/header/Header.hpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2010-2022 The Open-Transactions developers
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <boost/endian/buffers.hpp>
+#include <boost/endian/conversion.hpp>
```
Comment (TomaszAugustyn) [inline src/blockchain/bitcoin/block/header/Header.hpp]: Redundant.

File: src/blockchain/bitcoin/block/header/Header.hpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+#include <array>
```
Comment (TomaszAugustyn) [inline src/blockchain/bitcoin/block/header/Header.hpp]: Seems redundant.

File: src/blockchain/bitcoin/block/header/Header.hpp [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+#include <memory>
+
+#include "blockchain/block/header/Header.hpp"
+#include "blockchain/block/header/Imp.hpp"
+#include "internal/blockchain/bitcoin/block/Header.hpp"
+#include "internal/blockchain/block/Block.hpp"
+#include "opentxs/blockchain/BlockchainType.hpp"
+#include "opentxs/blockchain/Types.hpp"
+#include "opentxs/blockchain/bitcoin/NumericHash.hpp"
+#include "opentxs/blockchain/bitcoin/Work.hpp"
```
Comment (TomaszAugustyn) [inline src/blockchain/bitcoin/block/header/Header.hpp]: Redundant. Please revise all other includes as seems that a lot of them are not used.

File: src/blockchain/bitcoin/block/header/Header.hpp [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+#include "opentxs/core/Data.hpp"
+#include "opentxs/util/Bytes.hpp"
+#include "opentxs/util/Container.hpp"
+#include "opentxs/util/Numbers.hpp"
+#include "opentxs/util/Time.hpp"
+
+// NOLINTBEGIN(modernize-concat-nested-namespaces)
+namespace opentxs  // NOLINT
+{
+// inline namespace v1
```
Comment (TomaszAugustyn) [inline src/blockchain/bitcoin/block/header/Header.hpp]: What is it? Can we get rid of it?

File: src/blockchain/bitcoin/block/header/Header.hpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+class Header;
+}  // namespace block
+}  // namespace bitcoin
+
+namespace block
+{
+class Hash;
+}  // namespace block
+}  // namespace blockchain
+// }  // namespace v1
```
Comment (TomaszAugustyn) [inline src/blockchain/bitcoin/block/header/Header.hpp]: Also here.

File: generated/Version.hpp.in (hunk 1 above)
Comment (radexx99) [reply generated/Version.hpp.in:43]: MT-88 without todo in the code

File: src/blockchain/bitcoin/block/header/Header.hpp (hunk 5 above)
Comment (radexx99) [reply src/blockchain/bitcoin/block/header/Header.hpp:70]: MT-89

File: src/api/crypto/blockchain/Imp_blockchain.cpp (hunk 4 above)
Comment (radexx99) [reply src/api/crypto/blockchain/Imp_blockchain.cpp]: good point!

File: include/opentxs/blockchain/bitcoin/block/Header.hpp (hunk 2 above)
Comment (radexx99) [reply include/opentxs/blockchain/bitcoin/block/Header.hpp:30]: MT-89

**Pull request merged**